Automatically determine and setup dependencies

Matthimatiker opened the pull request:
Option to automatically detect associated entities and include these in the infrastructure setup (fixes #4).

Example:

    $infrastructure = ORMInfrastructure::createWithDependenciesFor(array(
        'My\Entity\ClassName'
    ));

Matthimatiker commented:
@MalteWunsch could you have a look at the code and check if there is room for improvement?

MalteWunsch commented:
Great code. Apart from my stubborn finals and privates, I've got nothing to add.

I'm wondering about the API:

* Why is the parameter of createWithDependenciesFor a string[]? I would've expected a simple singular string. My fear is that by enabling a user to create several independent dependency graphs, we'd encourage them to test too much in one test case.

* Why do we even have to provide a class as the starting point for the depedency graph? In contrast, we could automatically generate the whole Doctrine schema like with the Symfony command bin/console doctrine:schema:create. An argument for narrowing down the schema is that we reduce the complexity and hence isolate the tests better. A problem elsewhere in the Doctrine configuration wouldn't break existing tests. But is the whole schema generation not a valid approach, too? Clearly that's a separate issue, just wanted to share my thoughts here.

Matthimatiker commented:
> Why is the parameter of createWithDependenciesFor a string[]? 

Mainly because it was derived from our existing API :-)

> I would've expected a simple singular string. My fear is that by enabling a user to create several independent dependency graphs, we'd encourage them to test too much in one test case.

Point taken. It is already possible to pass just a single entity class. I kept the possibility to pass multiple entities (there might be valid use cases...), but I updated the documentation and changed the code examples to show the simple (and hopefully common) use case with a single entity.

> Why do we even have to provide a class as the starting point for the depedency graph? In contrast, we could automatically generate the whole Doctrine schema like with the Symfony command bin/console doctrine:schema:create.

Good point. To create the whole schema, some knowledge about the application and the directory layout is required. But it might be possible to create a bundle that simplifies the usage of the infrastructure in Symfony projects.


As you did not see any blocker, I will merge and tag this version.
